src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java

```diff
@@ -0,0 +1,130 @@
+     *
+     * Notice there are two apply() methods in this class:
+     * This apply() inherited from ReadWalker is marked final to discourage subclassing.
+     * A subclass must override the other apply() method that takes in the DuplicateSet.
+     */
+    @Override
+    public final void apply(GATKRead read, ReferenceContext referenceContext, FeatureContext featureContext) {
+        if (currentReadsWithSameUMI == null){ // evaluates to true for the very first read
+            currentReadsWithSameUMI = new ReadsWithSameUMI(read);
+            return;
+        }
+
+        final int readMoleculeNumber = MoleculeID.getMoleculeNumberOfRead(read);
+        final int duplicateSetMoleculeNumber = currentReadsWithSameUMI.getMoleculeNumber();
+
+        // If the incoming read has the molecule id less than that of the currentDuplicateSet,
+        // the input bam is not sorted properly by the MI tag
+        if (duplicateSetMoleculeNumber > readMoleculeNumber){
+            throw new UserException(String.format("The input bam must be sorted by the molecule ID (%s) tag.", SAMTag.MI.name()));
+        }
+
+        if (duplicateSetMoleculeNumber < readMoleculeNumber) {
+            // The incoming read's molecule ID does not match that of the current duplicate, meaning we've reached the end of the current set.
+            // Call the apply() method to process the current set and start a new set.
+            if (rejectSet(currentReadsWithSameUMI)){
+                currentReadsWithSameUMI = new ReadsWithSameUMI(read);
+                return;
+            }
+
+            apply(currentReadsWithSameUMI,
+                    new ReferenceContext(reference, currentReadsWithSameUMI.getInterval()), // Will create an empty ReferenceContext if reference or readInterval == null
+                    new FeatureContext(features, currentReadsWithSameUMI.getInterval()));
+            currentReadsWithSameUMI = new ReadsWithSameUMI(read);
+            return;
+        }
+
+        // The incoming read has the same UMI as the current set; simply add to the current set
+        currentReadsWithSameUMI.addRead(read);
+    }
+
+    /**
+     * A subclass must specify how to process the duplicate sets by overriding this method.
+     *
+     * @param readsWithSameUMI A set of reads with the matching UMIs with the same fragment start and end
+     * @param referenceContext A reference context object over the intervals determined by the duplicate set.
+     * @param featureContext Entries from a secondary feature file (e.g. vcf) if provided
+     *
+     */
+    public abstract void apply(ReadsWithSameUMI readsWithSameUMI, ReferenceContext referenceContext, FeatureContext featureContext );
+
+    /**
+     * A child tool class that override onTraversalSuccess() *must* call super.onTraversalSuccess() in order
```

Comment:
This is a bit inconsistent with the rest of the framework.  Lets consider changing this in a future refactor.

Reply:
Refactor done